src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.DecisionDagRewriter.cs

```diff
@@ -865,59 +870,232 @@ private void EnsureStringHashFunction(int labelsCount, SyntaxNode syntaxNode)
+                // So we can't just lower each `BoundWhenDecisionDagNode` separately, as that would result in duplicate blocks
+                // for the same `WhenExpression` and such expressions might contains labels which must be emitted once only.
+
+                // For a simple `BoundWhenDecisionDagNode` (with a unique `WhenExpression`), we lower to something like:
+                //   labelToSectionScope;
+                //   if (... logic from WhenExpression ...)
+                //   {
+                //     jump to whenTrue label
+                //   }
+                //   jump to whenFalse label
+
+                // For a complex `BoundWhenDecisionDagNode` (where the `WhenExpression` is shared), we lower to something like:
+                //   labelToSectionScope;
+                //   whenNodeIdentifierLocal = whenNodeIdentifier;
+                //   goto labelToWhenExpression;
+                //
+                // and we'll also create a section for the shared `WhenExpression` logic:
+                //   labelToWhenExpression;
+                //   if (... logic from WhenExpression ...)
+                //   {
+                //     jump to whenTrue label
+                //   }
+                //   switch on whenNodeIdentifierLocal with dispatches to whenFalse labels
+
+                // Prepared maps for `when` nodes and expressions
+                var whenExpressionMap = new Dictionary<BoundExpression, (LabelSymbol LabelToWhenExpression, List<BoundWhenDecisionDagNode> WhenNodes)>();
+                var whenNodeMap = new Dictionary<BoundWhenDecisionDagNode, (LabelSymbol LabelToWhenExpression, int WhenNodeIdentifier)>();
+                foreach (BoundDecisionDagNode node in sortedNodes)
+                {
+                    if (node is BoundWhenDecisionDagNode whenNode)
+                    {
+                        var whenExpression = whenNode.WhenExpression;
+                        if (whenExpression is not null && whenExpression.ConstantValue != ConstantValue.True)
+                        {
+                            LabelSymbol labelToWhenExpression;
+                            if (whenExpressionMap.TryGetValue(whenExpression, out var whenExpressionInfo))
+                            {
+                                labelToWhenExpression = whenExpressionInfo.LabelToWhenExpression;
+                                whenExpressionInfo.WhenNodes.Add(whenNode);
+                            }
+                            else
+                            {
+                                labelToWhenExpression = _factory.GenerateLabel("sharedWhenExpression");
+                                var list = new List<BoundWhenDecisionDagNode>();
+                                list.Add(whenNode);
+                                whenExpressionMap.Add(whenExpression, (labelToWhenExpression, list));
+                            }
+
+                            if (!whenNodeMap.ContainsKey(whenNode))
```

Comment:
>whenNodeMap.ContainsKey [](http://example.com/codeflow?start=33&length=23)

Are we sharing `BoundWhenDecisionDagNode` instances as well?

Reply:
No. You can have multiple `BoundWhenDecisionDagNode`s with same syntax, bindings, and when clause.

The relevant line is `state.Dag = uniqifyDagNode(new BoundWhenDecisionDagNode(first.Syntax, first.Bindings, first.WhenClause, whenTrue, whenFalse));` in `ComputeBoundDecisionDagNodes`.